File: BST/1. Insertion in a BST/python/iterative.py
# Install binarytree library for more flexibility: pip install binarytree
# from binarytree import Node

class Node:
    left = None
    right = None

    def __init__(self,value):
        self.value = value

def ascending(root):
    if root is None:
        return
    ascending(root.left)
    print(root.value, end=' ')
    ascending(root.right)

def iterativeInsert(root, key):
    if root is None:
        return Node(key)
    
    parent = None
    curr = root
    while (curr):
        parent = curr
        if key < curr.value:
            curr = curr.left
        else:
            curr = curr.right
        
    if key < parent.value:
        parent.left = Node(key)
    else:
        parent.right = Node(key)
    
    return root

keys = [15, 10, 20, 8, 12, 16, 25,30]

root = None
for key in keys:
    root = iterativeInsert(root,key)

ascending(root)
# print(root)     # Use this with imported Node class to print out tree data structure

File: BST/1. Insertion in a BST/Java/src/Iterative.java
public class Iterative {
    public static Node iterativeInsert(Node root, int key) {
        if (root == null) {
            return new Node(key);
        }

        Node curr = root;
        Node parent = new Node(0);

        while(curr != null) {
            parent = curr;
            if (key < curr.value) {
                curr = curr.left;
            }
            else {
                curr = curr.right;
            }
        }

        if (key < parent.value) {
            parent.left = new Node(key);
        }
        else {
            parent.right = new Node(key);
        }

        return root;
    }

    public static void main(String args[]) {
        Node root = null;

        int[] keys = {15, 10, 8, 20, 12, 16, 25, 30};

        for (int i = 0; i < keys.length; i++) {
            root = iterativeInsert(root, keys[i]);
        }

        root.ascending(root);
        
    }
}
